<h1>Analisi dati DataSet Hotel_Reviews</h1>

In [1]:
import pandas as pd
import numpy as np
import statistics as st

In [2]:
dfhotel = pd.read_csv("Hotel_Reviews.csv", decimal=".")

## EDA: Exploratory Data Analysis (Analisi Esplorativa dei Dati)

In [3]:
dfhotel.head()

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
0,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,8/3/2017,7.7,Hotel Arena,Russia,I am so angry that i made this post available...,397,1403,Only the park outside of the hotel was beauti...,11,7,2.9,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
1,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,8/3/2017,7.7,Hotel Arena,Ireland,No Negative,0,1403,No real complaints the hotel was great great ...,105,7,7.5,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
2,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/31/2017,7.7,Hotel Arena,Australia,Rooms are nice but for elderly a bit difficul...,42,1403,Location was good and staff were ok It is cut...,21,9,7.1,"[' Leisure trip ', ' Family with young childre...",3 days,52.360576,4.915968
3,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/31/2017,7.7,Hotel Arena,United Kingdom,My room was dirty and I was afraid to walk ba...,210,1403,Great location in nice surroundings the bar a...,26,1,3.8,"[' Leisure trip ', ' Solo traveler ', ' Duplex...",3 days,52.360576,4.915968
4,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/24/2017,7.7,Hotel Arena,New Zealand,You When I booked with your company on line y...,140,1403,Amazing location and building Romantic setting,8,3,6.7,"[' Leisure trip ', ' Couple ', ' Suite ', ' St...",10 days,52.360576,4.915968


In [4]:
dfhotel.tail()

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
515733,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/30/2015,8.1,Atlantis Hotel Vienna,Kuwait,no trolly or staff to help you take the lugga...,14,2823,location,2,8,7.0,"[' Leisure trip ', ' Family with older childre...",704 day,48.203745,16.335677
515734,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/22/2015,8.1,Atlantis Hotel Vienna,Estonia,The hotel looks like 3 but surely not 4,11,2823,Breakfast was ok and we got earlier check in,11,12,5.8,"[' Leisure trip ', ' Family with young childre...",712 day,48.203745,16.335677
515735,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/19/2015,8.1,Atlantis Hotel Vienna,Egypt,The ac was useless It was a hot week in vienn...,19,2823,No Positive,0,3,2.5,"[' Leisure trip ', ' Family with older childre...",715 day,48.203745,16.335677
515736,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/17/2015,8.1,Atlantis Hotel Vienna,Mexico,No Negative,0,2823,The rooms are enormous and really comfortable...,25,3,8.8,"[' Leisure trip ', ' Group ', ' Standard Tripl...",717 day,48.203745,16.335677
515737,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/9/2015,8.1,Atlantis Hotel Vienna,Hungary,I was in 3rd floor It didn t work Free Wife,13,2823,staff was very kind,6,1,8.3,"[' Leisure trip ', ' Family with young childre...",725 day,48.203745,16.335677


Visualizzo anche 5 righe casuali del DataSet così da vedere se, in modo casuale, riesco a trovare dei pattern nascosti

In [5]:
dfhotel.sample(5)

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
361091,La Rambla 22 24 Ciutat Vella 08002 Barcelona S...,223,3/23/2016,8.2,Eurostars Ramblas,Saudi Arabia,No Negative,0,1878,Excellent location,3,6,8.8,"[' Leisure trip ', ' Solo traveler ', ' Double...",498 day,41.378098,2.176201
291892,C rcega 368 Eixample 08037 Barcelona Spain,115,7/23/2016,8.4,Catalonia La Pedrera,United Kingdom,Pool is tiny Was photographed from a very fla...,12,1135,No Positive,0,1,8.3,"[' Leisure trip ', ' Family with older childre...",376 day,41.398609,2.162504
514645,Wrights Lane Kensington and Chelsea London W8 ...,1172,8/21/2016,7.8,Holiday Inn London Kensington,United Kingdom,The bed,3,5945,Food from room service,5,4,7.9,"[' Leisure trip ', ' Family with older childre...",347 day,51.499981,-0.192879
96487,20 Lavington Street Southwark London SE1 0NZ U...,804,8/26/2016,9.1,citizenM London Bankside,United Kingdom,I didn t like my length of stay I should have...,15,3802,We absolutely loved everything From the massi...,22,5,9.6,"[' Leisure trip ', ' Couple ', ' Double Room '...",342 day,51.505151,-0.100472
229600,66 Great Queen Street Covent Garden Camden Lon...,377,7/30/2017,8.5,Kingsway Hall Hotel,United Kingdom,Worst bed I ve ever tried to sleep in both to...,44,1363,Useful handheld device for sightseeing,7,1,5.0,"[' Leisure trip ', ' Couple ', ' Deluxe Double...",4 days,51.515864,-0.120063


#### Controllo dei tipi di dato
Verifico i tipi di dato delle variabili per assicurarmi che siano coerenti con l’analisi che voglio svolgere.

In [6]:
dfhotel.dtypes

Hotel_Address                                     str
Additional_Number_of_Scoring                    int64
Review_Date                                       str
Average_Score                                 float64
Hotel_Name                                        str
Reviewer_Nationality                              str
Negative_Review                                   str
Review_Total_Negative_Word_Counts               int64
Total_Number_of_Reviews                         int64
Positive_Review                                   str
Review_Total_Positive_Word_Counts               int64
Total_Number_of_Reviews_Reviewer_Has_Given      int64
Reviewer_Score                                float64
Tags                                              str
days_since_review                                 str
lat                                           float64
lng                                           float64
dtype: object

### Utilizzo il metodo describe()
Lo applico in due modalità:
- su tutto il DataFrame per avere una panoramica globale di tutte le variabili.
- solo sulle colonne Reviewer_Score e Average_Score, per analizzare nel dettaglio la distribuzione dei punteggi.

In [7]:
dfhotel.describe(include="all")

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
count,515738,515738.000000,515738,515738.000000,515738,515738,515738,515738.000000,515738.000000,515738,515738.000000,515738.000000,515738.000000,515738,515738,512470.000000,512470.000000
unique,1493,NaN,731,NaN,1492,227,330011,NaN,NaN,412601,NaN,NaN,NaN,55242,731,NaN,NaN
top,163 Marsh Wall Docklands Tower Hamlets London ...,NaN,8/2/2017,NaN,Britannia International Hotel Canary Wharf,United Kingdom,No Negative,NaN,NaN,No Positive,NaN,NaN,NaN,"[' Leisure trip ', ' Couple ', ' Double Room '...",1 days,NaN,NaN
freq,4789,NaN,2585,NaN,4789,245246,127890,NaN,NaN,35946,NaN,NaN,NaN,5101,2585,NaN,NaN
mean,NaN,498.081836,NaN,8.397487,NaN,NaN,NaN,18.539450,2743.743944,NaN,17.776458,7.166001,8.395077,NaN,NaN,49.442439,2.823803
std,NaN,500.538467,NaN,0.548048,NaN,NaN,NaN,29.690831,2317.464868,NaN,21.804185,11.040228,1.637856,NaN,NaN,3.466325,4.579425
min,NaN,1.000000,NaN,5.200000,NaN,NaN,NaN,0.000000,43.000000,NaN,0.000000,1.000000,2.500000,NaN,NaN,41.328376,-0.369758
25%,NaN,169.000000,NaN,8.100000,NaN,NaN,NaN,2.000000,1161.000000,NaN,5.000000,1.000000,7.500000,NaN,NaN,48.214662,-0.143372
50%,NaN,341.000000,NaN,8.400000,NaN,NaN,NaN,9.000000,2134.000000,NaN,11.000000,3.000000,8.800000,NaN,NaN,51.499981,0.010607
75%,NaN,660.000000,NaN,8.800000,NaN,NaN,NaN,23.000000,3613.000000,NaN,22.000000,8.000000,9.600000,NaN,NaN,51.516288,4.834443


In [8]:
dfhotel[["Reviewer_Score", "Average_Score"]].describe()

,Reviewer_Score,Average_Score
count,515738.000000,515738.000000
mean,8.395077,8.397487
std,1.637856,0.548048
min,2.500000,5.200000
25%,7.500000,8.100000
50%,8.800000,8.400000
75%,9.600000,8.800000
max,10.000000,9.800000


In [9]:
dfhotel.info()

<class 'pandas.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  str    
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  str    
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  str    
 5   Reviewer_Nationality                        515738 non-null  str    
 6   Negative_Review                             515738 non-null  str    
 7   Review_Total_Negative_Word_Counts           515738 non-null  int64  
 8   Total_Number_of_Reviews                     515738 non-null  int64  
 9   Positive_Review                             515738 non-null  str    
 10  Review_

In [10]:
dfhotel.shape

(515738, 17)

### Descrizione del DataFrame
Il DataFrame è composto da 17 colonne:

- Hotel_Address: variabile di tipo stringa che indica l’indirizzo dell’hotel.
- Additional_Number_of_Scoring: variabile di tipo int che indica il numero di punteggi validi associati a valutazioni prive di recensione testuale.
- Review_Date: variabile di tipo stringa che rappresenta la data di pubblicazione della recensione. È opportuno convertirla nel formato datetime per consentire analisi temporali.
- Average_Score: variabile di tipo float che indica il punteggio medio dell’hotel.
- Hotel_Name: variabile di tipo stringa che indica il nome dell’hotel.
- Reviewer_Nationality: variabile di tipo stringa che indica la nazionalità del recensore.
- Negative_Review: variabile di tipo stringa che contiene la recensione negativa dell’utente. In assenza di contenuto, il valore è “No negative”.
- Review_Total_Negative_Word_Counts: variabile di tipo int che indica il numero totale di parole presenti nella recensione negativa.
- Total_Number_of_Reviews: variabile di tipo int che indica il numero totale di recensioni ricevute dall’hotel.
- Positive_Review: variabile di tipo stringa che contiene la recensione positiva dell’utente. In assenza di contenuto, il valore è “No positive”.
- Review_Total_Positive_Word_Counts: variabile di tipo int che indica il numero totale di parole presenti nella recensione positiva.
- Total_Number_of_Reviews_Reviewer_Has_Given: variabile di tipo int che indica il numero totale di recensioni lasciate da un singolo utente.
- Reviewer_Score: variabile di tipo float che indica il punteggio assegnato all’hotel dall’utente.
- Tags: variabile di tipo stringa che contiene i tag assegnati dall’utente alla recensione dell’hotel.
- days_since_review: variabile di tipo stringa che indica il numero di giorni trascorsi tra la data della recensione e la data di raccolta dei dati. È opportuno convertirla in formato int per le successive analisi.
- lat: variabile di tipo float che indica la latitudine dell’hotel.
- lng: variabile di tipo float che indica la longitudine dell’hotel.

### Controllo se sono presenti valori NULL

In [11]:
dfhotel.isnull().sum()

Hotel_Address                                    0
Additional_Number_of_Scoring                     0
Review_Date                                      0
Average_Score                                    0
Hotel_Name                                       0
Reviewer_Nationality                             0
Negative_Review                                  0
Review_Total_Negative_Word_Counts                0
Total_Number_of_Reviews                          0
Positive_Review                                  0
Review_Total_Positive_Word_Counts                0
Total_Number_of_Reviews_Reviewer_Has_Given       0
Reviewer_Score                                   0
Tags                                             0
days_since_review                                0
lat                                           3268
lng                                           3268
dtype: int64

Dal controllo effettuato emerge che 3268 valori relativi alla latitudine e alla longitudine risultano mancanti. Questi casi rappresentano una piccola parte del dataset e saranno analizzati nella fase successiva di data cleaning per valutare la strategia più appropriata (rimozione o gestione dei valori mancanti).

### Controllo se sono presenti righe duplicate

In [12]:
dfhotel.duplicated().sum()

np.int64(526)

Dal controllo effettuato emerge la presenza di 526 righe duplicate. Tuttavia, è necessario verificare se si tratta di duplicati effettivi oppure di ripetizioni dovute alla struttura del dataset.

#### Verifica delle righe duplicate
Per analizzare la presenza di duplicati, seleziono tutte le righe duplicate, includendo anche le occorrenze originali.

In [13]:
dfhotel[dfhotel.duplicated(keep=False)].head()

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
23703,100 110 Euston Road Camden London NW1 2AJ Unit...,728,2/25/2017,8.9,Pullman London St Pancras,China,none,2,3168,good location and super nice staff room is bi...,13,39,9.6,"[' Business trip ', ' Solo traveler ', ' Class...",159 day,51.528677,-0.128349
23704,100 110 Euston Road Camden London NW1 2AJ Unit...,728,2/25/2017,8.9,Pullman London St Pancras,China,none,2,3168,good location and super nice staff room is bi...,13,39,9.6,"[' Business trip ', ' Solo traveler ', ' Class...",159 day,51.528677,-0.128349
61875,16 22 Great Russell Street Camden London WC1B ...,300,7/27/2017,9.0,The Bloomsbury Hotel,Israel,No Negative,0,1254,The attention received by Sebastian and his t...,12,4,9.6,"[' Leisure trip ', ' Couple ', ' Superior Doub...",7 days,51.517167,-0.129053
61876,16 22 Great Russell Street Camden London WC1B ...,300,7/27/2017,9.0,The Bloomsbury Hotel,Israel,No Negative,0,1254,The attention received by Sebastian and his t...,12,4,9.6,"[' Leisure trip ', ' Couple ', ' Superior Doub...",7 days,51.517167,-0.129053
68936,167 rue de Rome 17th arr 75017 Paris France,11,6/23/2017,6.8,Villa Eugenie,Netherlands,Location is very old not invested in past yea...,41,165,Location is central accessible by train parki...,12,24,4.6,"[' Business trip ', ' Solo traveler ', ' Singl...",41 days,48.887128,2.314205


#### Ordinamento dei duplicati per Hotel_Name e Rewiew_Date
Per verificarfe la struttura dei duplicati, le righe vengono ordinate per nome dell'hotel e data della recensione 

In [14]:
dfhotel[dfhotel.duplicated(keep=False)].sort_values(by=["Hotel_Name", "Review_Date"])

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
444764,Sch nbrunner Stra e 41 05 Margareten 1050 Vien...,144,2/23/2016,8.4,Arthotel ANA Prime,China,Price only includes room Everything else is e...,34,1813,Good location very clean and modern nice staff,10,26,7.5,"[' Leisure trip ', ' Couple ', ' Standard Doub...",527 day,48.192558,16.357607
444765,Sch nbrunner Stra e 41 05 Margareten 1050 Vien...,144,2/23/2016,8.4,Arthotel ANA Prime,China,Price only includes room Everything else is e...,34,1813,Good location very clean and modern nice staff,10,26,7.5,"[' Leisure trip ', ' Couple ', ' Standard Doub...",527 day,48.192558,16.357607
443175,Scarsdale Place Kensington Kensington and Chel...,1831,11/8/2016,8.1,Copthorne Tara Hotel London Kensington,United Kingdom,No Negative,0,7105,Location Breakfast,3,11,7.9,"[' Business trip ', ' Couple ', ' Standard Dou...",268 day,51.499046,-0.191707
443176,Scarsdale Place Kensington Kensington and Chel...,1831,11/8/2016,8.1,Copthorne Tara Hotel London Kensington,United Kingdom,No Negative,0,7105,Location Breakfast,3,11,7.9,"[' Business trip ', ' Couple ', ' Standard Dou...",268 day,51.499046,-0.191707
445413,Schepenbergweg 50 Zuidoost 1105 AT Amsterdam N...,315,8/31/2016,7.8,Fletcher Hotel Amsterdam,United Kingdom,No Negative,0,3773,Very nice hotel,5,6,10.0,"[' Leisure trip ', ' Couple ', ' Deluxe Double...",337 day,52.292457,4.944434
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
68997,167 rue de Rome 17th arr 75017 Paris France,11,8/5/2015,6.8,Villa Eugenie,Switzerland,nothing,2,165,the place was very clean and quite,8,1,7.5,"[' Leisure trip ', ' Family with older childre...",729 day,48.887128,2.314205
68951,167 rue de Rome 17th arr 75017 Paris France,11,9/19/2015,6.8,Villa Eugenie,Switzerland,Facilities did not function Sink was blocked ...,30,165,The staff tried,5,4,4.2,"[' Business trip ', ' Solo traveler ', ' Stand...",684 day,48.887128,2.314205
68982,167 rue de Rome 17th arr 75017 Paris France,11,9/19/2015,6.8,Villa Eugenie,Switzerland,Facilities did not function Sink was blocked ...,30,165,The staff tried,5,4,4.2,"[' Business trip ', ' Solo traveler ', ' Stand...",684 day,48.887128,2.314205
68961,167 rue de Rome 17th arr 75017 Paris France,11,9/22/2016,6.8,Villa Eugenie,Italy,Bed bugs air condition not work,7,165,Front office is helpfull,5,1,5.0,"[' Leisure trip ', ' Family with young childre...",315 day,48.887128,2.314205


#### Interpretazione risultati
L’analisi evidenzia la presenza di righe perfettamente identiche su tutte le variabili.

Tale situazione suggerisce la presenza di duplicati tecnici, probabilmente dovuti a errori di acquisizione o di merging del dataset.

Si valuta quindi la rimozione delle righe duplicate per evitare distorsioni nelle analisi successive.

## Data Cleaning

### Creazione copia del dataset
Per preservare l'integrità del dataset originale, viene creata una copia su cui vengono eseguite le operazioni di pulizia dati.

In [15]:
dfhotelClean = dfhotel.copy()

### Rimozione duplicati
In seguito all'analisi preliminare, si procede con l'eliminazione delle osservazioni duplicate che sono state identificate precedentemente nel dataset.

In [16]:
dfhotelClean = dfhotelClean.drop_duplicates()

### Standardizzazione dei nomi delle variabili
Per rendere il dataset più coerente e facilmente utilizzabile nelle successive analisi, in particolare in ambiente SQL, si procede alla standardizzazione dei nomi delle colonne.
In particolare, vengono rimosse maiuscole, in quanto il formato snake case è già adottato.

In [17]:
dfhotelClean.columns = dfhotelClean.columns.str.lower()

### Identificazione delle celle vuote nascoste (spazi)

Risulta necessario verificare la presenza di valori potenzialmente mancanti che non vengono riconosciuti come NaN.
In particolare, nei dataset importati da file CSV possono essere presenti celle contenenti solo spazi (" "), che non vengono interpretate automaticamente come valori nulli.

Per questo motivo, si procede a un controllo preliminare delle colonne testuali, al fine di individuare quante celle risultano vuote dopo la rimozione degli spazi iniziali e finali.

In [18]:
for col in dfhotelClean.select_dtypes(include=["object", "string"]):
    print(col, (dfhotelClean[col].str.strip() == "").sum())

hotel_address 0
review_date 0
hotel_name 0
reviewer_nationality 522
negative_review 849
positive_review 182
tags 0
days_since_review 0


Dopo aver identificato la presenza di celle contenenti solo spazi, è necessario procedere alla pulizia effettiva del dataset, in modo da:
- uniformare i dati testuali
- eliminare spazi iniziali e finali
- convertire i valori vuoti in NaN, rendendoli riconoscibili nelle analisi successive.

In [19]:
cols = dfhotelClean.select_dtypes(include=["object", "string"]).columns

for col in cols:
    dfhotelClean[col] = dfhotelClean[col].str.strip()

### Conversione delle stringhe vuote in valori mancanti

Dopo la rimozione degli spazi, le celle contenenti solo spazi diventano stringhe vuote (""), che devono essere convertite in valori mancanti (NaN).

In [20]:
dfhotelClean.replace("", np.nan, inplace=True)

,hotel_address,additional_number_of_scoring,review_date,average_score,hotel_name,reviewer_nationality,negative_review,review_total_negative_word_counts,total_number_of_reviews,positive_review,review_total_positive_word_counts,total_number_of_reviews_reviewer_has_given,reviewer_score,tags,days_since_review,lat,lng
0,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,8/3/2017,7.7,Hotel Arena,Russia,I am so angry that i made this post available ...,397,1403,Only the park outside of the hotel was beautiful,11,7,2.9,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
1,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,8/3/2017,7.7,Hotel Arena,Ireland,No Negative,0,1403,No real complaints the hotel was great great l...,105,7,7.5,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
2,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,7/31/2017,7.7,Hotel Arena,Australia,Rooms are nice but for elderly a bit difficult...,42,1403,Location was good and staff were ok It is cute...,21,9,7.1,"[' Leisure trip ', ' Family with young childre...",3 days,52.360576,4.915968
3,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,7/31/2017,7.7,Hotel Arena,United Kingdom,My room was dirty and I was afraid to walk bar...,210,1403,Great location in nice surroundings the bar an...,26,1,3.8,"[' Leisure trip ', ' Solo traveler ', ' Duplex...",3 days,52.360576,4.915968
4,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,7/24/2017,7.7,Hotel Arena,New Zealand,You When I booked with your company on line yo...,140,1403,Amazing location and building Romantic setting,8,3,6.7,"[' Leisure trip ', ' Couple ', ' Suite ', ' St...",10 days,52.360576,4.915968
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
515733,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/30/2015,8.1,Atlantis Hotel Vienna,Kuwait,no trolly or staff to help you take the luggag...,14,2823,location,2,8,7.0,"[' Leisure trip ', ' Family with older childre...",704 day,48.203745,16.335677
515734,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/22/2015,8.1,Atlantis Hotel Vienna,Estonia,The hotel looks like 3 but surely not 4,11,2823,Breakfast was ok and we got earlier check in,11,12,5.8,"[' Leisure trip ', ' Family with young childre...",712 day,48.203745,16.335677
515735,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/19/2015,8.1,Atlantis Hotel Vienna,Egypt,The ac was useless It was a hot week in vienna...,19,2823,No Positive,0,3,2.5,"[' Leisure trip ', ' Family with older childre...",715 day,48.203745,16.335677
515736,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/17/2015,8.1,Atlantis Hotel Vienna,Mexico,No Negative,0,2823,The rooms are enormous and really comfortable ...,25,3,8.8,"[' Leisure trip ', ' Group ', ' Standard Tripl...",717 day,48.203745,16.335677


### Gestione dei valori mancanti
Nella fase di analisi esplorativa (EDA) è stata evidenziata la presenza di valori mancanti all’interno del dataset.  
A seguito delle operazioni di pulizia preliminare (rimozione degli spazi e conversione delle stringhe vuote in NaN), si procede a un’analisi più approfondita della distribuzione dei valori mancanti, al fine di definire la strategia di trattamento più appropriata per ciascuna variabile.

In [21]:
dfhotelClean.isnull().sum()

hotel_address                                    0
additional_number_of_scoring                     0
review_date                                      0
average_score                                    0
hotel_name                                       0
reviewer_nationality                           522
negative_review                                849
review_total_negative_word_counts                0
total_number_of_reviews                          0
positive_review                                182
review_total_positive_word_counts                0
total_number_of_reviews_reviewer_has_given       0
reviewer_score                                   0
tags                                             0
days_since_review                                0
lat                                           3268
lng                                           3268
dtype: int64

Le variabili latitudine e longitudine presentano un numero limitato di valori mancanti (circa lo 0,6% del dataset).  
Considerata la ridotta incidenza, si è scelto di rimuovere le osservazioni incomplete al fine di garantire la qualità delle analisi successive.

In [22]:
dfhotelClean = dfhotelClean.dropna(subset=["lat", "lng"])

### Gestione positive_review e negative_review
Le variabili positive_review e negative_review contengono valori testuali standardizzati ("No Positive" e "No Negative") utilizzati per indicare l’assenza di una recensione rispettivamente positiva o negativa.

Tali valori non rappresentano dati mancanti, ma informazioni esplicite fornite dagli utenti.

Per garantire coerenza e uniformità, si procede alla normalizzazione del testo senza alterare il significato informativo delle osservazioni.

In [23]:
dfhotelClean["positive_review"] = dfhotelClean["positive_review"].str.lower()
dfhotelClean["negative_review"] = dfhotelClean["negative_review"].str.lower()


In [24]:
dfhotelClean[["positive_review", "negative_review"]].head(10)

,positive_review,negative_review
0,only the park outside of the hotel was beautiful,i am so angry that i made this post available ...
1,no real complaints the hotel was great great l...,no negative
2,location was good and staff were ok it is cute...,rooms are nice but for elderly a bit difficult...
3,great location in nice surroundings the bar an...,my room was dirty and i was afraid to walk bar...
4,amazing location and building romantic setting,you when i booked with your company on line yo...
5,good restaurant with modern design great chill...,backyard of the hotel is total mess shouldn t ...
6,the room is spacious and bright the hotel is l...,cleaner did not change our sheet and duvet eve...
7,good location set in a lovely park friendly st...,apart from the price for the brekfast everythi...
8,no positive,even though the pictures show very clean rooms...
9,the room was big enough and the bed is good th...,the aircondition makes so much noise and its h...


A seguito delle operazioni di pulizia (rimozione degli spazi, conversione delle stringhe vuote in valori mancanti e normalizzazione del testo), si analizzano i valori mancanti nelle variabili positive_review e negative_review.

In [25]:
dfhotelClean[["positive_review", "negative_review"]].isnull().sum()

positive_review    180
negative_review    835
dtype: int64

Dall’analisi emerge la presenza di un numero ridotto di valori mancanti:
- positive_review: 180 valori mancanti
- negative_review: 835 valori mancanti

Considerata la dimensione complessiva del dataset (circa 515.000 osservazioni), l’incidenza dei valori mancanti risulta trascurabile.


In conclusione, le variabili di positive_review e negative_review presentano tre situazioni distinte:
- recensioni testuali valide
- assenza esplicita di feedback (“no positive”, “no negative”)
- valori mancanti residui (NaN)

Considerata la bassa incidenza dei valori mancanti nelle variabili, si è scelto di non procedere all’eliminazione delle osservazioni.
La decisione è motivata dalla natura descrittiva delle variabili, che non risultano critiche per la struttura complessiva del dataset.

### Gestione reviewer_nationality
La variabile reviewer_nationality presenta valori mancanti a seguito delle operazioni di pulizia del dataset.
In particolare, questi valori non derivano da errori strutturali, ma da informazioni non esplicitate dagli utenti.
Per evitare la perdita di osservazioni e mantenere la variabile utilizzabile nelle analisi descrittive, i valori mancanti vengono sostituiti con una categoria esplicita ("unknonw").

In [26]:
dfhotelClean["reviewer_nationality"] = dfhotelClean["reviewer_nationality"].fillna("unknown")

In [27]:
dfhotelClean.isnull().sum()

hotel_address                                   0
additional_number_of_scoring                    0
review_date                                     0
average_score                                   0
hotel_name                                      0
reviewer_nationality                            0
negative_review                               835
review_total_negative_word_counts               0
total_number_of_reviews                         0
positive_review                               180
review_total_positive_word_counts               0
total_number_of_reviews_reviewer_has_given      0
reviewer_score                                  0
tags                                            0
days_since_review                               0
lat                                             0
lng                                             0
dtype: int64

### Conversione della variabile temporale

La variabile review_date deve essere convertita nel formato datetime al fine di consentire analisi temporali più strutturate, come l’estrazione di informazioni relative ad anno e mese.

In [28]:
dfhotelClean["review_date"] = pd.to_datetime(dfhotelClean["review_date"])

In [29]:
dfhotelClean["review_date"].dtype

dtype('<M8[us]')

### Estrazione di variabili temporali
A partire dalla variabile review_date vengono estratti l’anno e il mese della recensione, utili per l'analisi.

In [30]:
dfhotelClean["year"] = dfhotelClean["review_date"].dt.year
dfhotelClean["month"] = dfhotelClean["review_date"].dt.month

In [31]:
dfhotelClean.head()

,hotel_address,additional_number_of_scoring,review_date,average_score,hotel_name,reviewer_nationality,negative_review,review_total_negative_word_counts,total_number_of_reviews,positive_review,review_total_positive_word_counts,total_number_of_reviews_reviewer_has_given,reviewer_score,tags,days_since_review,lat,lng,year,month
0,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,2017-08-03,7.7,Hotel Arena,Russia,i am so angry that i made this post available ...,397,1403,only the park outside of the hotel was beautiful,11,7,2.9,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968,2017,8
1,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,2017-08-03,7.7,Hotel Arena,Ireland,no negative,0,1403,no real complaints the hotel was great great l...,105,7,7.5,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968,2017,8
2,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,2017-07-31,7.7,Hotel Arena,Australia,rooms are nice but for elderly a bit difficult...,42,1403,location was good and staff were ok it is cute...,21,9,7.1,"[' Leisure trip ', ' Family with young childre...",3 days,52.360576,4.915968,2017,7
3,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,2017-07-31,7.7,Hotel Arena,United Kingdom,my room was dirty and i was afraid to walk bar...,210,1403,great location in nice surroundings the bar an...,26,1,3.8,"[' Leisure trip ', ' Solo traveler ', ' Duplex...",3 days,52.360576,4.915968,2017,7
4,s Gravesandestraat 55 Oost 1092 AA Amsterdam N...,194,2017-07-24,7.7,Hotel Arena,New Zealand,you when i booked with your company on line yo...,140,1403,amazing location and building romantic setting,8,3,6.7,"[' Leisure trip ', ' Couple ', ' Suite ', ' St...",10 days,52.360576,4.915968,2017,7


### Conversione della variabile days_since_review
La variabile days_since_review è attualmente rappresentata come stringa e contiene informazioni nel formato "X days".

Per poterla utilizzare nelle analisi quantitative, si procede all’estrazione della componente numerica e alla conversione in formato intero. Per una sicurezza maggiore, elimino anche eventuali spazi iniziali e finali.

In [32]:
dfhotelClean["days_since_review"] = dfhotelClean["days_since_review"].str.strip().str.split().str[0].astype(int)

In [33]:
dfhotelClean["days_since_review"].head()

0     0
1     0
2     3
3     3
4    10
Name: days_since_review, dtype: int64

In [34]:
dfhotelClean["days_since_review"].dtypes

dtype('int64')

### Creazione di Variabili Derivate (Feature Engineering)

In questa fase, vengono generate nuove variabili a partire da quelle esistenti per arricchire l'analisi. 
Verranno create tre variabili derivate:
- review_length, calcolata sommando il numero di parole delle recensioni positive e negative, così da misurare l'engagement totale dell'utente. Sarà utile per verificare se esiste una correlazione tra la loquacità del cliente e il punteggio assegnato.
- hotel_country, viene estratta la nazione dalla colonna hotel_address così da permettere di aggregare i risultati e confrontare le medie dei punteggi tra i diversi paesi europei rappresentati nel dataset.
- hotel_city, viene estratta dalla colonna hotel_address, utile per confrontare le performance degli hotel tra le diverse città europee presenti nel dataset.

In [35]:
dfhotelClean["review_length"] = dfhotelClean["review_total_negative_word_counts"] + dfhotelClean["review_total_positive_word_counts"]
dfhotelClean[["review_total_negative_word_counts","review_total_positive_word_counts","review_length"]].head()

,review_total_negative_word_counts,review_total_positive_word_counts,review_length
0,397,11,408
1,0,105,105
2,42,21,63
3,210,26,236
4,140,8,148


In [36]:
dfhotelClean["hotel_country"] = dfhotelClean["hotel_address"].str.strip().str.split().str[-1]
dfhotelClean['hotel_country'] = dfhotelClean['hotel_country'].replace('Kingdom', 'United Kingdom')
dfhotelClean["hotel_country"].sample(10)

104183    United Kingdom
222560    United Kingdom
378028           Austria
320048       Netherlands
386287       Netherlands
219132    United Kingdom
29904     United Kingdom
117104            France
256653            France
406042             Italy
Name: hotel_country, dtype: object

Al fine di comprendere la struttura della variabile hotel_address e individuare la posizione della città all’interno della stringa, si procede con un’analisi esplorativa degli indirizzi unici. In particolare, vengono estratte le ultime due componenti di ciascun indirizzo, ipotizzando che possano contenere informazioni rilevanti come città e paese.

In [37]:
indirizzi_unici = dfhotelClean["hotel_address"].unique()
esplorazione = pd.Series(indirizzi_unici).str.split().str[-2:]
print(esplorazione.value_counts().head(20))

[Paris, France]             455
[United, Kingdom]           399
[Barcelona, Spain]          208
[Milan, Italy]              162
[Vienna, Austria]           147
[Amsterdam, Netherlands]    105
Name: count, dtype: int64


In [38]:
citta_uK = dfhotelClean[dfhotelClean["hotel_address"].str.contains("United Kingdom")]
citta_uK["hotel_address"].str.split().str[-5:].head(10)

405    [London, SW5, 9NB, United, Kingdom]
406    [London, SW5, 9NB, United, Kingdom]
407    [London, SW5, 9NB, United, Kingdom]
408    [London, SW5, 9NB, United, Kingdom]
409    [London, SW5, 9NB, United, Kingdom]
410    [London, SW5, 9NB, United, Kingdom]
411    [London, SW5, 9NB, United, Kingdom]
412    [London, SW5, 9NB, United, Kingdom]
413    [London, SW5, 9NB, United, Kingdom]
414    [London, SW5, 9NB, United, Kingdom]
Name: hotel_address, dtype: object

Dall'analisi emerge che il dataset è circoscritto a sei città europee: Londra, Parigi, Amsterdam, Vienna, Barcellona e Milano.

Poiché la struttura delle stringhe varia tra i diversi paesi, rendendo inaffidabile un'estrazione basata solo sulla posizione, si è deciso di utilizzare una funzione di mappatura testuale. Questo garantisce di associare correttamente ogni hotel alla propria città, indipendentemente dal numero di parole o dai codici postali presenti nell'indirizzo.

In [39]:
lista_citta = ["London", "Paris", "Barcelona", "Milan", "Vienna", "Amsterdam"]

def trova_citta(indirizzo):
    for citta in lista_citta:
        if citta in indirizzo:
            return citta
    return "Altra città"

dfhotelClean["hotel_city"] = dfhotelClean["hotel_address"].apply(trova_citta)
dfhotelClean["hotel_city"].sample(10)

324873       Vienna
110121       London
226566       London
274678    Amsterdam
121552        Paris
345407       London
442733       London
121507        Paris
471305    Barcelona
197150       London
Name: hotel_city, dtype: str

In [40]:
dfhotelClean.sample(5)

,hotel_address,additional_number_of_scoring,review_date,average_score,hotel_name,reviewer_nationality,negative_review,review_total_negative_word_counts,total_number_of_reviews,positive_review,...,reviewer_score,tags,days_since_review,lat,lng,year,month,review_length,hotel_country,hotel_city
499202,Villarroel 106 Eixample 08011 Barcelona Spain,235,2015-09-08,7.8,Acevi Villarroel,Canada,i was there on business and only upon arriving...,153,2251,no positive,...,6.7,"[' Business trip ', ' Family with young childr...",695,41.385378,2.156723,2015,9,153,Spain,Barcelona
312698,Delflandlaan 15 Slotervaart 1062 EA Amsterdam ...,947,2015-10-27,8.7,Best Western Premier Hotel Couture,United Kingdom,no matters is any significance,6,8177,nice room good staff v good breakfast very rea...,...,8.8,"[' Leisure trip ', ' Solo traveler ', ' Standa...",646,52.351114,4.841163,2015,10,17,Netherlands,Amsterdam
35388,110 rue Blomet 15th arr 75015 Paris France,39,2016-03-23,8.5,Hotel Eden,Canada,wish we didn t have to sign in to wifi every t...,19,477,the staff went above and beyond to make my fam...,...,9.2,"[' Leisure trip ', ' Family with young childre...",498,48.840561,2.299104,2016,3,41,France,Paris
430735,Robert Stolz Platz 1 01 Innere Stadt 1010 Vien...,299,2016-12-27,8.3,Le Meridien Vienna,Iran,very croudy it was difficult even to talk with...,22,2393,no positive,...,5.0,"[' Leisure trip ', ' Family with young childre...",219,48.202730,16.366121,2016,12,22,Austria,Vienna
236387,7 Pepys Street City of London London EC3N 4AF ...,1936,2017-02-27,8.7,DoubleTree by Hilton Hotel London Tower of London,United Kingdom,staff attentive but could smile more without w...,41,7491,good location impressive lounge bar and comfy ...,...,6.3,"[' Leisure trip ', ' Group ', ' King Room ', '...",157,51.510841,-0.078058,2017,2,50,United Kingdom,London


### Esportazione del dataset pulito

Al termine della fase di data cleaning, il dataset è stato esportato in formato CSV, generando il file `hotel_reviews_clean.csv`, utilizzato nelle successive fasi di analisi.

In [41]:
dfhotelClean.to_csv("hotel_reviews_clean.csv", index=False)